### ISSUE TITLE
Music Search and Play Capability ready for testing

### ISSUE THREAD

**Issue by jplourde5** [jishi/node-sonos-http-api#258]
The README has been updated to explain how it works. Supports Apple Music, Spotify, Deezer, and the local Sonos music library. You can get the releases at:

node-sones-http-api Use the below release instead of what is is specified in the documentation
https://github.com/jplourde5/node-sonos-http-api/releases

For the Amazon Echo users, the release also includes default room capabilities...

echo-sonos 
https://github.com/jplourde5/echo-sonos/releases

**Comment by ppitkin** [comment]
Thanks for all the great work. I’ll try to get a test system build to try this out.

 

One question I have is how to I ask for the system (echo or API) to play a specific album from an specific artist. I never select individual tracks – only albums.

 

Peter

**Comment by jplourde5** [comment]
Use 'album' type in the URL instead of the 'song' type. The README provides some instructions on the options.  I haven't enabled the album type for echo-Sonos yet.  I'm concerned about there being too many "play" utterances, and I'm thinking about what might make sense as a different trigger utterance. Maybe 'spin album' ...  Are you trying to use echo-Sonos?

**Comment by ppitkin** [comment]
OK so the API supports it – that’s fine (sorry I must have missed that in the ReadMe L). I will simply refactor your Alex skill to remove the track utterances and replace them with something suitable for an album.

It’s been a good week so far: you guys making such fantastic progress with the Sonos API and Alexa skill;  and the news from Sonos that they now support Plex (so it’s possible to now stream everything from my Plex server which will also automatically do on the fly format conversions).

 

Once I get that all that running my next task will be to extend the  Node-Sonos-HTTP-Api so that it controls my Russound in parallel to the Sonos devices.

In my setup I don’t have Sonos units in each room – they weren’t available when I built my chalet and I had installed a very expensive Russound system with 16 zones plus sub-zones. I currently have 4 Sonos “connect” units feeding 4 of the 8 input sources (Once I get my hands on an Amazon Dot I will have that connected as a source as well J).

The idea is to use the Sonos skill to say what to play in what room and then set the relevant Sonos connect to that stream and turn the Russound on for that room.

Deciding which Sonos to use for the room source will be the tricky part. Setting the Sonos volume will be redirected to the Russound controller for that room. It’s all a bit of a hack but the Russound cost me so much money, but it’s very good, so I don’t want to throw it away.

 

So the idea is to effectively combine the two action:

·         Alexa, tell Sonos to play Thriller by Michael Jackson on Russound input 1

·         Alexa, tell Russound to place input 1 in the Living Room

Into one action:

·         Alexa, tell Sonos to play Thriller by Michael Jackson in the Living Room

**Comment by jplourde5** [comment]
Easy addition.  Even the Lambda code has it in there.  Make these simple changes:

Add this to your Intents

    {
      "intent": "AlbumIntent",
      "slots": [
        {
            "name": "Name",
            "type": "AMAZON.LITERAL"
        },
        {
            "name": "Room",
            "type": "ROOMS"
        }   
      ]
    },
      

Add these to your utterances

AlbumIntent play album {boston|Name} in the {Room}
AlbumIntent play album {full circle|Name} in the {Room}
AlbumIntent play album {stranger to stranger|Name} in the {Room}
AlbumIntent play album {a cure for loneliness|Name} in the {Room}
AlbumIntent play album {a sailor's guide to earth|Name} in the {Room}
AlbumIntent play album {you will never be one of us|Name} in the {Room}
AlbumIntent play album {for those of you that have never|Name} in the {Room}

AlbumIntent play album {boston|Name}
AlbumIntent play album {full circle|Name}
AlbumIntent play album {stranger to stranger|Name}
AlbumIntent play album {a cure for loneliness|Name}
AlbumIntent play album {a sailor's guide to earth|Name}
AlbumIntent play album {you will never be one of us|Name}
AlbumIntent play album {for those of you that have never|Name}

**Comment by jplourde5** [comment]
...and let me know if you have any trouble.  

The money spent on the Russound is water under the bridge.  The real question is will you have a better solution if you just switch to using the Sonos?  Constantly having to switch the Russound input sounds like a pain to me.  But maybe it has some great features.  Or maybe it simply turns into your amp and speaker system ;)  I wired my entire house for a central system many years ago when I built it.  I bought the Sonos ZP 80 (=Connect) the day they first came out, hooked it up to my central system, and have used no other input since.  Little by little, room by room, I've brought in independent Sonos components (I think I have more than 10 now), and the rooms that I like the least are the ones that are still on the central solution :)

**Comment by jplourde5** [comment]
Added the album intent back into the distribution.  Tested and it seems to work fine.

**Comment by jplourde5** [comment]
Just an FYI.  The node-sones-http-api and echo-sonos combination for my test distributions are working for all testers.  Attention needs to be paid to making sure that the Alexa Skill Invocation Word is all lowercase.  Example, use 'sonos', not 'Sonos' for the invocation word when setting up the Alexa skill.

**Comment by thosirl** [comment]
Just grabbed latest jplourde5 beta branch and getting following error when trying /Kitchen/musicsearch/deezer/song/albatross (and otheres)

`Error: Internal Server Error
    at ClientRequest.<anonymous> (/home/thos/node-sonos-http-api-1.0.0-beta.17.12/node_modules/sonos-discovery/lib/helpers/request.js:22:21)
    at ClientRequest.g (events.js:260:16)
    at emitOne (events.js:77:13)
    at ClientRequest.emit (events.js:169:7)
    at HTTPParser.parserOnIncomingClient (_http_client.js:430:21)
    at HTTPParser.parserOnHeadersComplete (_http_common.js:103:23)
    at Socket.socketOnData (_http_client.js:320:20)
    at emitOne (events.js:77:13)
    at Socket.emit (events.js:169:7)
    at readableAddChunk (_stream_readable.js:153:18)
    at Socket.Readable.push (_stream_readable.js:111:10)
    at TCP.onread (net.js:536:20)`

Where can I see more debug info? Anything I can provide to help?

**Comment by thosirl** [comment]
PS - coming from a working jishi install, other play/pause function works fine, can see all zones etc.

**Comment by thosirl** [comment]
Using library returns {"status":"error"}  and 'undefined' in the console logs

Using /kitchen/musicsearch/deezer/artist/nirvana also just returns same error and undefined.

**Comment by jplourde5** [comment]
Do you have Deezer standard or Elite?  Use elite instead of deezer if you have Elite.

**Comment by thosirl** [comment]
Nope, Deezer 'plain' here.

**Comment by jplourde5** [comment]
Make sure that you have the latest version. I tested the library and it did act a little sketchy.  I'll take a look in the morning.

**Comment by jplourde5** [comment]
I found and fixed the problem with the library search.  Single word searches worked fine but must-word searches were failing due encoding characters.  I ran tests with both deezer and elite, and both continue to work fine.  Make sure that you have the latest releases or standard Deezer will not work and it will fail with the error that you are seeng.

node-sones-http-api Use the below release instead of what is is specified in the documentation
https://github.com/jplourde5/node-sonos-http-api/releases
(currently jplourde5/node-sonos-http-api-v1.0.0-beta.17.13)

For the Amazon Echo users, the release also includes default room capabilities...

echo-sonos 
https://github.com/jplourde5/echo-sonos/releases
(currently jplourde5/echo-sonos-Beta.8)

**Comment by thosirl** [comment]
Turns out I've got a different Deezer account type, which is '519' and not '43015', so changing serviceTypes (line 238) and serviceTokens (line 273), and I'm up and running.
Checking with another Sonos/Deezer buddy to see what his is. I'm in Ireland, any chance they regionalize these?
Llibrary working too, and Echo integration looking good, thank you!

**Comment by jplourde5** [comment]
I added some capability in the latest release to detect and save a different code for a different region.  Simply add a song type from the desired service and issue  

.../ROOM_NAME/musicsearch/SERVICE/setsid/NAME_OF_SONG  

Where 

ROOM_NAME can be any valid room
SERVICE  can be apple, spotify, seeker, or elite  (deeper in your case)
NAME_OF SONG should be the name of the song that you saved in Sonos Favorites exactly as it is spelled in favorites

The function will inspect the favorite to grab the valid sid, update the value for the specified service, and store it on disk where it will be loaded and used from there on out

**Comment by jishi** [comment]
What is the difference between deezer and elite?

**Comment by jplourde5** [comment]
Not much. Elite uses FLAC which changes some of the metadata, mainly in the uri.

**Comment by jishi** [comment]
But we don't know which subscription the user has, is that correct? No credentials is required to access the API either?

**Comment by jplourde5** [comment]
There is no way to auto-detect it since just about everything, including the sid, are the same for both.  The deezer service directive uses standard mp3 and elite uses FLAC.  Standard works for both types of accounts.  FLAC blows up if you send it to a standard Deezer account.  We could just send everything the standard mp3 route but that would mean that someone would not be able to take advantage of FLAC, which over here in the US, may be the main reason for buying Deezer.

It would be great if there was some way to auto-detect and simply have the deezerDef use FLAC or mp3, depending on the account type, but I can't see a way to do that.  Let me know if you have some ideas.  I'm trying to avoid having to use settings for this, but it is another option.  A third option might be to have another command type for Deezer that turns on FLAC if someone wants it and the setting could be saved on disk.  So right now I see 3 options for fully supporting both options:

1. Use separate service names (what is being done now)
2. Store a Deezer FLAC option in settings that would have to be adjusted by hand
3. Default to mp3 and create a deezer type command 'setflac' to set and save a setting
4. Any other idea that you can come up with

**Comment by jplourde5** [comment]
Eliminated 99.9% of the contents of eliteDef.  eliteDef equals deezerDef with an init call that sets the song uri to the FLAC setting.

**Comment by thosirl** [comment]
Music search working well at present, via Echo, very happy. Thank you!

One thing I'm noticing however is that for 'play pearl jam' , I get the same song order each time (Jeremy, Even Flow ...), this is the order Deezer seem to be returning the search.
I tried enabling shuffle at this point:
```
                .then(() => player.coordinator.addURIToQueue(tracks.queueTracks[0].uri, tracks.queueTracks[0].metadata, true, 1))
                .then(() => player.coordinator.shuffle(true))
                .then(() => player.coordinator.play())
```
but because the tracks are still being added async, it's ineffective at actually shuffling (ie; it shuffles a playlist of 1, then other tracks are added in sequence), hence I still get Jeremy, Even Flow etc track order.

For 'Artist' request, I feel a shuffle or random order would give better experience, as opposed to an 'Album' request, when track order may be preferred behaviour.

I can see the Deezer API , has an 'order' parameter, but doesnt seem to do a shuffle/randomizer ...

Any thoughts? Will 'proper' artist/station results from Deezer, or same?

**Comment by jplourde5** [comment]
Me and others have the same wish, but I'm not sure how to pull it off unless I wait for the queue to get loaded before it starts playing, which can take a while.  The problem is as you observed, in that the order that the services are sending is the same and the solution purposely starts playing as soon as the first song is loaded, which is always the same.  What follows should be random.  I've thought about waiting for N number of tracks to load before playing the first song, and at least it will be somewhat different from time to time, but always one of the first N songs.  The larger N is, the longer it will be before music starts playing. 

Open to ideas...

**Comment by jplourde5** [comment]
Added a call to shuffle the track list so that now you get a different first track each time when playing tracks for an artist.

https://github.com/jplourde5/node-sonos-http-api/releases/tag/v1.0.0-beta.17.16

**Comment by GregRocket** [comment]
@jplourde5 I've installed and setup beta.17.16. I only have a sonos music library that is very large. When I issue a command such as /office/musicsearch/library/album/{search term} nothing plays. If I look at a console of the api it shows "loading library"

**Comment by jplourde5** [comment]
It may take a while to load the first time.  It caches the library list to disk and loads it up on startup, which should go faster after the first time.  How many tracks do you have?

**Comment by GregRocket** [comment]
A little over 50k tracks. Does the cache stay on the disk if the server is exited? If new songs are added or removed does the cache just update new songs? I'm running on a Raspberry PI 3, any idea how long it might take? I do have a mounted 32gig sd card so I don't think space should be an issue. Where is the cache file list directory?
Thanks,
Greg

**Comment by GregRocket** [comment]
@jplourde5 I ran office/musicsearch/library/load for a couple of hours until the server finally stopped or locked up.

**Comment by jplourde5** [comment]
You should see a file called library.json getting generated in the api lib folder.  I only have about 2k tracks, so I'm not going to be able to troubleshoot this for you.  There are two different areas that may be going sideways, one being the generation of the library file and the second being the in-memory solution for the fuzzy search. 

1. Are you seeing the library.json file, and if so, how large is it?
2. How much memory is available for the app?

**Comment by GregRocket** [comment]
When I run office/musicsearch/library/load the console shows:
Loading Library
Loading Library
Loading Library
When I look in the directory api lib folder I don't see a file library.json getting generated.

The PI 3 has 1GB of memory. When I run free -m this is the results:
             total       used       free     shared    buffers     cached
Mem:     973        239        733         12         26        140

**Comment by jplourde5** [comment]
Grab and replace the libraryDef.js file that I just updated to provide some more clues in the console and let me know what you see.

In my jplourde5 Beta
node-sonos-http-api/lib/music_services/libraryDef.js

**Comment by GregRocket** [comment]
@jplourde5 
OK it's been running for a while. Here is a snapshot of console output:

Loading Library
Track Count 0
Track Count 1000
Track Count 2000
Track Count 3000
Track Count 4000
Track Count 5000
Track Count 6000
Loading Library
Track Count 0
Track Count 7000
Track Count 1000
Track Count 8000
Track Count 2000
Track Count 9000
Track Count 3000
Loading Library
Track Count 0
Track Count 10000
Track Count 4000
Track Count 1000
Track Count 11000
Track Count 5000
Track Count 2000
Loading Library
Track Count 0
Track Count 12000
Track Count 6000
Track Count 3000
Track Count 1000
Track Count 13000
Track Count 7000
Loading Library
Track Count 0
Track Count 4000
Track Count 2000
Track Count 14000
Track Count 8000
Track Count 1000
Track Count 5000
Loading Library
Track Count 0
Track Count 3000
Track Count 15000
Track Count 9000
Track Count 2000

**Comment by jplourde5** [comment]
Interesting.  Shows the Load Library function being called multiple times while it is still running.  Are you making API calls while the library is loading?  If so, don't invoke the API again until the Loading Library stops.

Regardless, I will also add a check to see if it is running already and not execute again if it is,

**Comment by GregRocket** [comment]
I am not making any calls to the api. Just letting the load command.

**Comment by jplourde5** [comment]
I just added the fix to keep it from running multiple times.  Download and update the libraryDef.js which may fix the problem if you aren't doing something that is invoking the API while the load is running.

**Comment by GregRocket** [comment]
@jplourde5 OK here is the output:

Loading Library
Track Count 0
Track Count 1000
Track Count 2000
Track Count 3000
Track Count 4000
Track Count 5000
Track Count 6000
/root/node-sonos-http-api-beta/lib/music_services/libraryDef.js:128
    return resolve('Loading');
           ^

ReferenceError: resolve is not defined
    at loadLibrary (/root/node-sonos-http-api-beta/lib/music_services/libraryDef.js:128:12)
    at Object.loadLibrarySearch [as load] (/root/node-sonos-http-api-beta/lib/music_services/libraryDef.js:177:12)
    at Object.musicSearch (/root/node-sonos-http-api-beta/lib/actions/musicSearch.js:251:23)
    at handleAction (/root/node-sonos-http-api-beta/lib/sonos-http-api.js:102:35)
    at HttpAPI.requireDir.requestHandler.handleAction.then [as requestHandler] (/root/node-sonos-http-api-beta/lib/sonos-http-api.js:79:5)
    at /root/node-sonos-http-api-beta/server.js:80:13
    at Server.finish (/root/node-sonos-http-api-beta/node_modules/node-static/lib/node-static.js:111:13)
    at finish (/root/node-sonos-http-api-beta/node_modules/node-static/lib/node-static.js:170:14)
    at /root/node-sonos-http-api-beta/node_modules/node-static/lib/node-static.js:144:17
    at FSReqWrap.oncomplete (fs.js:82:15)

npm ERR! Linux 4.4.9-v7+
npm ERR! argv "/usr/local/bin/node" "/usr/local/bin/npm" "start"
npm ERR! node v4.3.2
npm ERR! npm  v3.9.0
npm ERR! code ELIFECYCLE
npm ERR! sonos-http-api@1.0.0-beta.17 start: `node server.js`
npm ERR! Exit status 1
npm ERR!
npm ERR! Failed at the sonos-http-api@1.0.0-beta.17 start script 'node server.js'.
npm ERR! Make sure you have the latest version of node.js and npm installed.
npm ERR! If you do, this is most likely a problem with the sonos-http-api package,
npm ERR! not with npm itself.
npm ERR! Tell the author that this fails on your system:
npm ERR!     node server.js
npm ERR! You can get information on how to open an issue for this project with:
npm ERR!     npm bugs sonos-http-api
npm ERR! Or if that isn't available, you can get their info via:
npm ERR!     npm owner ls sonos-http-api
npm ERR! There is likely additional logging output above.

npm ERR! Please include the following file with any support request:
npm ERR!     /root/node-sonos-http-api-beta/npm-debug.log

**Comment by jplourde5** [comment]
Grab the libraryDef.js again or just change line 128 to 

return Promise.resolve('Loading');

**Comment by GregRocket** [comment]
@jplourde5 OK now the library.json file is created. 

I ran office/musicsearch/library/album/david+bowie+heroes and after a short wait the album started to play.

Looks good now. I'll run some more tests to be sure. Thanks!

**Comment by GregRocket** [comment]
@jplourde5 This is odd. musicsearch was working fine last night. I did shutdown the server overnight and restarted this morning. I made no changes. Now when I try something like office/musicsearch/library/album/david+bowie+heroes I keep getting an error. Other commands work correctly just not musicsearch.

Error: Internal Server Error
    at ClientRequest.<anonymous> (/root/node-sonos-http-api-beta/node_modules/sonos-discovery/lib/helpers/request.js:22:21)
    at ClientRequest.g (events.js:260:16)
    at emitOne (events.js:77:13)
    at ClientRequest.emit (events.js:169:7)
    at HTTPParser.parserOnIncomingClient (_http_client.js:412:21)
    at HTTPParser.parserOnHeadersComplete (_http_common.js:88:23)
    at Socket.socketOnData (_http_client.js:302:20)
    at emitOne (events.js:77:13)
    at Socket.emit (events.js:169:7)
    at readableAddChunk (_stream_readable.js:146:16)
    at Socket.Readable.push (_stream_readable.js:110:10)
    at TCP.onread (net.js:523:20)

Any idea what could be causing this now?

**Comment by thosirl** [comment]
Is your sonos-http-api process still running?

**Comment by jplourde5** [comment]
I would think that it would need to be running or he wouldn't be getting that error.  How long did you wait after the restart before you started issuing commands?  I'm wondering if is taking some time to load the library.json file.

**Comment by GregRocket** [comment]
I only stopped the API. I restarted it using npm start. It's been running for a couple of hours now and still getting error.

**Comment by jplourde5** [comment]
I normally see that error when there is something wrong with the metadata.  Have you tried some albums that worked before?  Do individual songs work?  Artists?  

Fishing to see if it just a problem with the album type or maybe even a specific album.

**Comment by GregRocket** [comment]
OK here is the problem. I had office grouped with other rooms. This is when the errors occur. If I ungroup office by itself, then office/musicsearch/library/album/david+bowie+heroes and other searches work.

**Comment by jplourde5** [comment]
Interesting.  I may be able to reproduce that.  Just for added information, does the problem occur with both albums and songs, or just albums?

**Comment by GregRocket** [comment]
I've discovered some issues. 
**Album:**
If any of the rooms player has [no music] and is ungrouped then the error occurs.  Any ungrouped room that's player has [no music], then the error occurs

If the first room (coordinator?) player, in a group of rooms, has [no music] then the error occurs. A group of rooms where the first room's player has [no music] then error. 

If the groups first room's player has something in the player and other rooms have [no music] then it works correctly.

When it works correctly the album is added to the sonos queue and starts playing.

Has something to do with [no music] in the player.

**Songs:**
Seems to be same issue as with albums. If any ungrouped room or group of rooms first room has [no music] does not work correctly. Works if ungrouped room or first room in group has something in the music player, not [no music]

**Comment by jplourde5** [comment]
Great info.  This will help a great deal.  I'll try to fix it later this evening.

**Comment by jplourde5** [comment]
Well, I thought that this would be easy.  I'm not seeing the issue with the empty queue. I've tried grouping and ungrouping.  When you see the error, if you go to the Sonos App, do you see anything in the queue?  Do you see anything in the Playing window, or does it say no music?

**Comment by GregRocket** [comment]
![1](https://cloud.githubusercontent.com/assets/12052578/18109530/b6fb0f8a-6ec6-11e6-8046-aa40d370d65d.jpg)
If Room Player looks like Office, or Office is the first room in a group of rooms, then does not work.

![2](https://cloud.githubusercontent.com/assets/12052578/18109532/ba543af8-6ec6-11e6-8c9c-883bdba0e189.jpg)
If Room Player looks like Hall Bathroom or Hall Bathroom is the first room in a group of rooms, then it works.

**Comment by jplourde5** [comment]
Thanks.  When playing an ALBUM, AFTER the error, does the Sonos app look the same or does it show a song?  

Just trying to get a feel for if the error happens at the point of queuing the first song, or after.

Also, if you play just one SONG, do you still get the error?

Playing one song uses some different logic and I'm trying to see if both logic branches experience the same issue.

This info will point me to where to add some logging for more clues.

**Comment by GregRocket** [comment]
Album:

Command:
/media room/musicsearch/library/album/david+bowie+heroes

Error:
{"status":"error","error":"Error: Internal Server Error\n at ClientRequest.<anonymous> (/root/node-sonos-http-api-beta/node_modules/sonos-discovery/lib/helpers/request.js:22:21)\n at ClientRequest.g (events.js:260:16)\n at emitOne (events.js:77:13)\n at ClientRequest.emit (events.js:169:7)\n at HTTPParser.parserOnIncomingClient (_http_client.js:412:21)\n at HTTPParser.parserOnHeadersComplete (_http_common.js:88:23)\n at Socket.socketOnData (_http_client.js:302:20)\n at emitOne (events.js:77:13)\n at Socket.emit (events.js:169:7)\n at readableAddChunk (_stream_readable.js:146:16)\n at Socket.Readable.push (_stream_readable.js:110:10)\n at TCP.onread (net.js:523:20)"}

Result:
![sonos_2016-08-30_17-57-57](https://cloud.githubusercontent.com/assets/12052578/18112402/2922e7da-6edc-11e6-8b6c-1f459a99cdfb.png)

Song:
Same result

**Comment by jplourde5** [comment]
So it looks like the first song is getting queued and then it fails.  Grab and install the musicSearch.js file that I just updated and we should get more clues.  I also removed the shuffle on an album type since I suspect most would want an album played in order.  Shouldn't be part of the problem because it is not used for a song type.  Shoot me your console output after a failure.

**Comment by GregRocket** [comment]
More strange behavior. If I select Office ungrouped, then album search and play works correctly. It does not matter if the sonos player shows [no music] or not for Office. Maybe because it is the last room alphabetically? Office also works if it is the first selected in a group of rooms, even though office shows [no music] 

**Note:** All eleven songs for the album loaded into the queue and start playing.

**Command**:
/office/musicsearch/library/album/journey,%20evolution

**Console Output**:
uri x-file-cifs://Mybooklive/mp3/Albums/Journey/Journey%20-%20Discography/1979%20Evolution/07%20ovin'%20You%20Is%20Easy.mp3

met <DIDL-Lite xmlns:dc="http://purl.org/dc/elements/1.1/" xmlns:upnp="urn:schemas-upnp-org:metaata-1-0/upnp/"
          xmlns:r="urn:schemas-rinconnetworks-com:metadata-1-0/" xmlns="urn:schemas-upnp-org:metdata-1-0/DIDL-Lite/">
          <item id="song" parentID="undefinedS://Mybooklive/mp3/Albums/Journey/Journey%20-%20Disography/1979%20Evolution/07%20Lovin'%20You%20Is%20Easy.mp3" restricted="true"><dc:title>Journey/volution</dc:title><upnp:class>object.undefined</upnp:class>
          <desc id="cdudn" nameSpace="urn:schemas-rinconnetworks-com:metadata-1-0/">RINCON_AssocatedZPUDN</desc></item></DIDL-Lite>

1
3
4
5



**-------------------------------**

Now any other room that the sonos player shows [no music] does not work and comes up with error. In this test it is for media room.

**Note:** Only one song for the album loaded into the queue, song 7 of 11 songs, and the queue did not start playing.

**Command:**
/media%20room/musicsearch/library/album/journey,%20evolution

**Console Output:**
uri x-file-cifs://Mybooklive/mp3/Albums/Journey/Journey%20-%20Discography/1979%20Evolution/07%20Lovin'%20You%20Is%20Easy.mp3

met <DIDL-Lite xmlns:dc="http://purl.org/dc/elements/1.1/" xmlns:upnp="urn:schemas-upnp-org:metadata-1-0/upnp/"
          xmlns:r="urn:schemas-rinconnetworks-com:metadata-1-0/" xmlns="urn:schemas-upnp-org:metadata-1-0/DIDL-Lite/">
          <item id="song" parentID="undefinedS://Mybooklive/mp3/Albums/Journey/Journey%20-%20Discography/1979%20Evolution/07%20Lovin'%20You%20Is%20Easy.mp3" restricted="true"><dc:title>Journey/Evolution</dc:title><upnp:class>object.undefined</upnp:class>
          <desc id="cdudn" nameSpace="urn:schemas-rinconnetworks-com:metadata-1-0/">RINCON_AssociatedZPUDN</desc></item></DIDL-Lite>

1
3
Error: Internal Server Error
    at ClientRequest.<anonymous> (/root/node-sonos-http-api-beta/node_modules/sonos-discovery/lib/helpers/request.js:22:21)
    at ClientRequest.g (events.js:260:16)
    at emitOne (events.js:77:13)
    at ClientRequest.emit (events.js:169:7)
    at HTTPParser.parserOnIncomingClient (_http_client.js:412:21)
    at HTTPParser.parserOnHeadersComplete (_http_common.js:88:23)
    at Socket.socketOnData (_http_client.js:302:20)
    at emitOne (events.js:77:13)
    at Socket.emit (events.js:169:7)
    at readableAddChunk (_stream_readable.js:146:16)
    at Socket.Readable.push (_stream_readable.js:110:10)
    at TCP.onread (net.js:523:20)

**----------------------------**

For this test I started a music streaming service in the media room. This time it works correctly with no error.

**Note:** All eleven songs for the album loaded into the queue and start playing.

**Command:**
/media%20room/musicsearch/library/album/journey,%20evolution

**Console:**
uri x-file-cifs://Mybooklive/mp3/Albums/Journey/Journey%20-%20Discography/1979%20Evolution/07%20Lovin'%20You%20Is%20Easy.mp3

met <DIDL-Lite xmlns:dc="http://purl.org/dc/elements/1.1/" xmlns:upnp="urn:schemas-upnp-org:metadata-1-0/upnp/"
          xmlns:r="urn:schemas-rinconnetworks-com:metadata-1-0/" xmlns="urn:schemas-upnp-org:metadata-1-0/DIDL-Lite/">
          <item id="song" parentID="undefinedS://Mybooklive/mp3/Albums/Journey/Journey%20-%20Discography/1979%20Evolution/07%20Lovin'%20You%20Is%20Easy.mp3" restricted="true"><dc:title>Journey/Evolution</dc:title><upnp:class>object.undefined</upnp:class>
          <desc id="cdudn" nameSpace="urn:schemas-rinconnetworks-com:metadata-1-0/">RINCON_AssociatedZPUDN</desc></item></DIDL-Lite>

1
2
4
5

**Comment by jplourde5** [comment]
Ah, lose the comma and see what happens

**Comment by GregRocket** [comment]
Well OK sorry for all the trouble. Seems to be working now.

**Comment by jplourde5** [comment]
No problem.  I'm happy that you have it all working now.

**Comment by jishi** [comment]
This is merged into master now. Let's open new issues if we find stuff that needs fixing.
